stlab/channel.hpp

```diff
@@ -505,12 +627,207 @@ struct shared_process : shared_process_receiver<yield_type<T, Arg>>,
+
+template <std::size_t i, typename C>
+struct context_worker
+{
+    std::shared_ptr<C> _context;
+
+    explicit context_worker(const std::shared_ptr<C>& c) : _context(c) {}
+
+    // auto as return type does not work here with VS 2015 Update 3. 
+    // The trait inside get_process_state does not see it, grmpf
+    std::pair<process_state, std::chrono::system_clock::time_point> state() const { 
+        auto s = _context->template state<i>(); 
+        return s;
+    }
+    
+    template <typename U>
+    void await(U&& u) { _context->template await<i,U>(std::forward<U>(u)); }
+
+    auto yield() { return _context->yield(); }
+
+    void close() { _context->close(); }
+};
+
+
+template <std::size_t I, typename C, typename R>
+auto create_receiver_collector(std::shared_ptr<C>& context, R&& upstream_receiver) {
+    auto receiver_collector = std::forward<R>(upstream_receiver) | buffer_size(1) | context_worker<I, C>(context);
+    return receiver_collector._p;
+}
+
+template <typename C, typename... R, std::size_t... I>
+auto create_receiver_collectors(std::shared_ptr<C>& context, std::index_sequence<I...>, R&&... upstream_receiver) {
+    return std::make_tuple(create_receiver_collector<I>(context, std::forward<R>(upstream_receiver))...);
+}
+
+template <typename C, typename S, typename P, typename RC, std::size_t... I>
+auto create_joined_process(S&& s, P&& p, RC& receiver_collectors, std::index_sequence<I...>) {
+    return std::make_shared<detail::shared_process<P, std::shared_ptr<C>>>(
+        std::forward<S>(s), std::forward<P>(p), std::get<I>(receiver_collectors)...);
+}
+
+template <typename C, typename RC, typename P, std::size_t... I>
+void map_as_sender_(RC& receiver_collectors, P& p, std::index_sequence<I...>) {
+    (void)std::initializer_list<int>{(std::get<I>(receiver_collectors)->map(sender<std::shared_ptr<C>>(p)), 0)...};
+}
+
+template <typename C, typename RC, typename P>
+void map_as_sender(RC& receiver_collectors, P& p) {
+    map_as_sender_<C>(receiver_collectors, p, std::make_index_sequence<std::tuple_size<RC>::value>());
+}
+
+template <typename F, typename C, std::size_t... I>
+auto apply_join_args_(const F& f, std::shared_ptr<C>& context, std::index_sequence<I...>) {
+    return f(std::move(std::get<I>(context->_result)).value()...);
+}
+
+template <typename F, typename C>
+auto apply_join_args(const F& f, std::shared_ptr<C>& context) {
+    return apply_join_args_(f, context, std::make_index_sequence<std::tuple_size<decltype(context->_result)>::value>());
+}
+
+
+template <typename R, typename F, typename C>
+struct join_processor
+{
+    F                   _f;
+    std::shared_ptr<C>  _context;
+    bool                _closed = false;
+    explicit join_processor(F&& f) : _f(std::move(f)) {}
+
+    void await(const std::shared_ptr<C>& context) {
+        // FP check here the ref counts, to handle a closed upstream ?
+        _context = context;
+    }
+
+    R yield() {
+        auto wc = make_weak_ptr(_context);
```

Comment:
The SCOPE_EXIT here seems unnecessary. c is going to fall out of scope releasing the context. This code looks like the equivalent of:
```
return detail::apply_join_args(_f, std::move(_context));
```
Or:
```
auto hold = std::move(_context);
return detail::apply_join_args(_f, hold);
```